### PR TITLE
Fix execve

### LINKED ISSUE #2624: Possible issues with stack
### Bug Description

It seems that whenever we invoke mirrord exec, we are running out of stack, and mirrord is silently disappearing.

In this case, we are running ruby version 3.2.1 on linux. 

Can confirm this doesnt happen on mac. 

Can confirm 3.111.0 does not exhibit this issue

### Steps to Reproduce

1. Update to 3.112.0
2. Run processes that has lots of childs with mirrord exec
3. Observe mirrord disappear into the ether...

### Backtrace
```shell
SocketError: Failed to open TCP connection to risk.default.svc.cluster.local:80 (getaddrinfo: Temporary failure in name resolution)
            test/helpers/exemplar.rb:78:in `get_request'
```

### Relevant Logs
```shell
...
2024-07-31T18:27:51.426372Z TRACE ThreadId(01) mirrord_layer::socket::hooks: hooked "dup2"
2024-07-31T18:27:51.426381Z TRACE ThreadId(01) mirrord_layer::socket::hooks: hooked "getpeername"
2024-07-31T18:27:51.426390Z TRACE ThreadId(01) mirrord_layer::socket::hooks: hooked "getsockname"
2024-07-31T18:27:51.426398Z TRACE ThreadId(01) mirrord_layer::socket::hooks: hooked "gethostname"
2024-07-31T18:27:51.426409Z TRACE ThreadId(01) mirrord_layer::socket::hooks: hooked "accept4"
2024-07-31T18:27:51.426418Z TRACE ThreadId(01) mirrord_layer::socket::hooks: hooked "dup3"
2024-07-31T18:27:51.426425Z TRACE ThreadId(01) mirrord_layer::socket::hooks: hooked "accept"
2024-07-31T18:27:51.426437Z TRACE ThreadId(01) mirrord_layer::socket::hooks: hooked "gethostbyname"
2024-07-31T18:27:51.426445Z TRACE ThreadId(01) mirrord_layer::socket::hooks: hooked "getaddrinfo"
2024-07-31T18:27:51.426459Z TRACE ThreadId(01) mirrord_layer::socket::hooks: hooked "freeaddrinfo"
2024-07-31T18:27:51.426468Z TRACE ThreadId(01) mirrord_layer::exec_hooks::hooks: hooked "execv"
2024-07-31T18:27:51.426477Z TRACE ThreadId(01) mirrord_layer::exec_hooks::hooks: hooked "execve"
2024-07-31T18:27:51.426695Z  INFO ThreadId(01) mirrord_layer: Initializing mirrord-layer!
2024-07-31T18:27:51.426726Z DEBUG ThreadId(01) mirrord_layer: Loaded into executable executable=Some("/usr/bin/diff") args=Some(ExecuteArgs { exec_name: "diff", invoked_as: "diff", args: ["diff", "-u", "/tmp/expect20240731-1113580-sp8nfx", "/tmp/butwas20240731-1113580-fa2ke0"] }) pid=1113859 parent_pid=1113580 env_vars=Vars { 
...
("RAILS_ENV", "test"), ("RACK_ENV", "development")] }
```

### Your operating system and version

Linux 6.8.5 kernel

### Local process

bundle 

### Local process version

ruby 3.21. bundler 2.4.6

### Additional Info

Unfortunately no bypass on socket logs :
```
...
2024-07-31T18:53:07.029355Z TRACE ThreadId(01) mirrord_layer::socket::ops: getaddrinfo -> result 0x00000000043c2ec0
2024-07-31T18:53:07.029449Z TRACE ThreadId(01) mirrord_layer::socket::ops: in connect LazyLock(
2024-07-31T18:53:07.029908Z TRACE ThreadId(01) mirrord_layer::socket::ops: we are connected Connected {
2024-07-31T18:53:07.029964Z TRACE ThreadId(01) mirrord_layer::socket::ops: getsockname -> local_address Ip(
2024-07-31T18:53:07.138523Z TRACE ThreadId(01) mirrord_layer::socket::ops: getaddrinfo -> result 0x0000000004458b00
2024-07-31T18:53:07.138609Z TRACE ThreadId(01) mirrord_layer::socket::ops: in connect LazyLock(
2024-07-31T18:53:07.139432Z TRACE ThreadId(01) mirrord_layer::socket::ops: we are connected Connected {
2024-07-31T18:53:07.139506Z TRACE ThreadId(01) mirrord_layer::socket::ops: getsockname -> local_address Ip(
...
```

### PR DESCRIPTION
**Pull request by aviramha**
Tested (manually):

- VSCode e2e
- Ruby
- Node

Closes #2624

### CODE REVIEW

**Comment by aviramha** [comment]
The node integration fails only with node 14 (doesn't happen on Node 22).

**Comment by aviramha** [comment]
I'm going mental
adding 
```rs
eprintln!("execveeee");
```
to `execve_detour` seem to make the integration test work :|

**Comment by eyalb181** [comment]
Does the number of "e"s at the end change the outcome

**Comment by aviramha** [comment]
if I add
```rs
    if std::process::id() == 83333 {
        eprintln!("asdasd");
    }
```
It still gets stuck, which leads me to believe it's not stack but the actual call to eprintln!
maybe it's a proper node issue?

**Comment by aviramha** [comment]
Okay it happens with 22 as well :|

**Comment by aviramha** [comment]
also `println` makes it go away..

**Comment by aviramha** [comment]
I added a loop on AtomicBool and attached a debugger. inspected all variables by hand and saw exec happening to the process as expected. GOD PLEASE HELP ME

**Comment by aviramha** [comment]
if I check the `syscall` file of the process, it seems to be stuck on `gettid`:
```
cat /proc/95565/syscall 
220 0x4111 0xffffeec104e0 0x0 0x0 0xffffa56ee640 0xfffffffe7fffffff 0xffffeec104e0 0xffffa3b230d0
```

**Comment by aviramha** [comment]
okay, appearntly it uses a flag called vfork that means that the child needs to call execve or exit before it returns, hence why we can't attach
I found the child, then attached to it and found out we're stuck on our own mutex!

**Comment by aviramha** [comment]
Ready for review!

File: mirrord/layer/src/socket/hooks.rs [hunk 1]
```diff
@@ -267,7 +271,10 @@ unsafe extern "C" fn getaddrinfo_detour(
     getaddrinfo(rawish_node, rawish_service, rawish_hints)
         .map(|c_addr_info_ptr| {
             out_addr_info.copy_from_nonoverlapping(&c_addr_info_ptr, 1);
-            MANAGED_ADDRINFO.insert(c_addr_info_ptr as usize);
+            MANAGED_ADDRINFO
+                .lock()
+                .expect("couldn't lock")
```
Comment (DmitryDodzin) [inline mirrord/layer/src/socket/hooks.rs:276]:
```suggestion
                .expect("MANAGED_ADDRINFO lock failed")
```

File: mirrord/layer/src/socket/hooks.rs [hunk 2]
```diff
@@ -295,29 +302,27 @@ unsafe extern "C" fn getaddrinfo_detour(
...
-            }
-        })
-        .unwrap_or_else(|| {
-            // If the `addrinfo` pointer was not allocated by `getaddrinfo_detour`, then it
-            // is bypassed.
-            FN_FREEADDRINFO(addrinfo);
-        })
+    let mut managed_addr_info = MANAGED_ADDRINFO
+        .lock()
+        .expect("couldn't lock MANAGED_ADDRINFO");
```
Comment (DmitryDodzin) [inline mirrord/layer/src/socket/hooks.rs:307]:
```suggestion
        .expect("MANAGED_ADDRINFO lock failed");
```

**Review by DmitryDodzin — APPROVED** [review]
Only Nitpicks on consistent expect messages

**Pull request merged**